File: MiniDOMOffice/CRM/Configuration/CCRMConfig.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;

namespace minidom
{
    public partial class CustomerCalls
    {
        /// <summary>
        /// Flag della configurazione del CRM
        /// </summary>
        [Flags]
        public enum CRMFlags : int
        {
            /// <summary>
            /// Nessun flag
            /// </summary>
            None = 0,

            /// <summary>
            /// Segnala le chiamate senza risposta ai supervisori
            /// </summary>
            SEGNALA_NONRISPONDE = 1,

            /// <summary>
            /// Segnala le chiamate fatte a persone la cui data di ricontatto non é prossima
            /// </summary>
            SEGNALA_RICONTATTITROPPOLONTANI = 2,

            /// <summary>
            /// Inva ai supervisori le statistiche giornaliere sugli operatori
            /// </summary>
            INVIA_STATISTICHE_GIORNALIERE = 4
        }

        /// <summary>
        /// Classe che rappresenta la configurazione del CRM
        /// </summary>
        /// <remarks></remarks>
        public class CCRMConfig 
            : DMD.XML.DMDBaseXMLObject 
        {
            private string m_FromAddress;
            private string m_FromDisplayName;
            private CRMFlags m_Flags;
            private string m_TargetAddress;
            private int m_MinutiRicontatto;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CCRMConfig()
            {
                m_FromAddress = "";
                m_FromDisplayName = "";
                m_Flags = CRMFlags.None;
                m_TargetAddress = "";
                m_MinutiRicontatto = 15;
            }

            /// <summary>
            /// Restituisce o imposta un valore booleano che indica se il sistema deve segnalare le telefonate senza risposta
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool SegnalaNonRisponde
            {
                get
                {
                    return DMD.RunTime.TestFlag(m_Flags, CRMFlags.SEGNALA_NONRISPONDE);
                }

                set
                {
                    if (SegnalaNonRisponde == value)
                        return;
                    m_Flags = DMD.RunTime.SetFlag(m_Flags, CRMFlags.SEGNALA_NONRISPONDE, value);
                    DoChanged("SegnalaNonRisponde", value, !value);
                }
            }

            /// <summary>
            /// Restituisce o imposta un valore booleano che indica se il sistema deve segnalare la programmazioni du un ricontatto oltre il limite di MinutiRicontatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool SegnalaRicontattiTroppoLontani
            {
                get
                {
                    return DMD.RunTime.TestFlag(m_Flags, CRMFlags.SEGNALA_RICONTATTITROPPOLONTANI);
                }

                set
                {
                    if (SegnalaRicontattiTroppoLontani == value)
                        return;
                    m_Flags = DMD.RunTime.SetFlag(m_Flags, CRMFlags.SEGNALA_RICONTATTITROPPOLONTANI, value);
                    DoChanged("SegnalaRicontattiTroppoLontani", value, !value);
                }
            }


            /// <summary>
            /// Restituisce o imposta un valore booleano che indica se il sistema deve segnalare la programmazioni du un ricontatto oltre il limite di MinutiRicontatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool InviaStatisticheGiornaliere
            {
                get
                {
                    return DMD.RunTime.TestFlag(m_Flags, CRMFlags.INVIA_STATISTICHE_GIORNALIERE);
                }

                set
                {
                    if (InviaStatisticheGiornaliere == value)
                        return;
                    m_Flags = DMD.RunTime.SetFlag(m_Flags, CRMFlags.INVIA_STATISTICHE_GIORNALIERE, value);
                    DoChanged("InviaStatisticheGiornaliere", value, !value);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'indirizzo e-mail utilizzato per inviare le statistiche e gli alert del CRM
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string FromAddress
            {
                get
                {
                    return m_FromAddress;
                }

                set
                {
                    string oldValue = m_FromAddress;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_FromAddress = value;
                    DoChanged("FromAddress", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta il nome visualizzato come mittente
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string FromDisplayName
            {
                get
                {
                    return m_FromDisplayName;
                }

                set
                {
                    string oldValue = m_FromDisplayName;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_FromDisplayName = value;
                    DoChanged("FromDisplayName", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'indirizzo e-mail a cui inviare le segnalazioni e le statistiche
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string TargetAddress
            {
                get
                {
                    return m_TargetAddress;
                }

                set
                {
                    string oldValue = m_TargetAddress;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_TargetAddress = value;
                    DoChanged("TargetAddress ", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta flags aggiuntivi
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public CRMFlags Flags
            {
                get
                {
                    return m_Flags;
                }

                set
                {
                    var oldValue = m_Flags;
                    if (oldValue == value)
                        return;
                    m_Flags = value;
                    DoChanged("Flags", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta i minuti oltre i quali il nuovo ricontatto programmato deve essere segnalato
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public int MinutiRicontatto
            {
                get
                {
                    return m_MinutiRicontatto;
                }

                set
                {
                    int oldValue = m_MinutiRicontatto;
                    if (oldValue == value)
                        return;
                    m_MinutiRicontatto = value;
                    DoChanged("MinutiRicontatto", value, oldValue);
                }
            }

          

            //protected override bool LoadFromRecordset(DBReader reader)
            //{
            //    this.m_FromAddress = reader.Read("FromAddress", this.m_FromAddress);
            //    this.m_FromDisplayName = reader.Read("FromDisplayName", this.m_FromDisplayName);
            //    this.m_Flags = reader.Read("Flags", this.m_Flags);
            //    this.m_TargetAddress = reader.Read("TargetAddress", this.m_TargetAddress);
            //    this.m_MinutiRicontatto = reader.Read("MinutiRicontatto", this.m_MinutiRicontatto);
            //    return base.LoadFromRecordset(reader);
            //}

            //protected override bool SaveToRecordset(DBWriter writer)
            //{
            //    writer.Write("FromAddress", m_FromAddress);
            //    writer.Write("FromDisplayName", m_FromDisplayName);
            //    writer.Write("Flags", m_Flags);
            //    writer.Write("TargetAddress", m_TargetAddress);
            //    writer.Write("MinutiRicontatto", m_MinutiRicontatto);
            //    return base.SaveToRecordset(writer);
            //}

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("FromAddress", m_FromAddress);
                writer.WriteAttribute("FromDisplayName", m_FromDisplayName);
                writer.WriteAttribute("Flags", (int?)m_Flags);
                writer.WriteAttribute("TargetAddress", m_TargetAddress);
                writer.WriteAttribute("MinutiRicontatto", m_MinutiRicontatto);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "FromAddress":
                        {
                            m_FromAddress = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "FromDisplayName":
                        {
                            m_FromDisplayName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Flags":
                        {
                            m_Flags = (CRMFlags)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "TargetAddress":
                        {
                            m_TargetAddress = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "MinutiRicontatto":
                        {
                            m_MinutiRicontatto = DMD.Integers.ValueOf(DMD.XML.Utils.Serializer.DeserializeString(fieldValue));
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Carica la configurazione dalle impostazioni di sistema
            /// </summary>
            public void Load()
            {
                //string dbSQL = "SELECT * FROM [" + GetTableName() + "] ORDER BY [ID] ASC";
                //var dbRis = GetConnection().ExecuteReader(dbSQL);
                //if (dbRis.Read())
                //{
                //    GetConnection().Load(this, dbRis);
                //}

                //dbRis.Dispose();
                var str = minidom.Sistema.ApplicationContext.Settings.GetItemByKey("CRM.Configuration", "");
                var tmp = DMD.XML.Utils.Deserialize<CCRMConfig>(str);
                if (tmp is object)
                    DMD.RunTime.CopyFrom(this, tmp, false);
            }

            /// <summary>
            /// Salva la configurazione
            /// </summary>
            public void Save()
            {
                //base.Save();
                var str = DMD.XML.Utils.Serialize(this);
                minidom.Sistema.ApplicationContext.Settings.SetItemByKey("CRM.Configuration", str);
                minidom.CustomerCalls.CRM.SetConfig(this);
            }
        }
    }
}

File: MiniDOMOffice/CRM/CRM/CCRMConfig.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases

Imports minidom.Anagrafica



Partial Public Class CustomerCalls

    <Flags> _
    Public Enum CRMFlags As Integer
        None = 0


        SEGNALA_NONRISPONDE = 1
        SEGNALA_RICONTATTITROPPOLONTANI = 2
        INVIA_STATISTICHE_GIORNALIERE = 4
    End Enum

    ''' <summary>
    ''' Classe che rappresenta la configurazione del CRM
    ''' </summary>
    ''' <remarks></remarks>
    Public Class CCRMConfig
        Inherits DBObject

        Private m_FromAddress As String
        Private m_FromDisplayName As String
        Private m_Flags As CRMFlags
        Private m_TargetAddress As String
        Private m_MinutiRicontatto As Integer

        Public Sub New()
            Me.m_FromAddress = ""
            Me.m_FromDisplayName = ""
            Me.m_Flags = CRMFlags.None
            Me.m_TargetAddress = ""
            Me.m_MinutiRicontatto = 15
        End Sub

        ''' <summary>
        ''' Restituisce o imposta un valore booleano che indica se il sistema deve segnalare le telefonate senza risposta
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property SegnalaNonRisponde As Boolean
            Get
                Return TestFlag(Me.m_Flags, CRMFlags.SEGNALA_NONRISPONDE)
            End Get
            Set(value As Boolean)
                If (Me.SegnalaNonRisponde = value) Then Exit Property
                Me.m_Flags = SetFlag(Me.m_Flags, CRMFlags.SEGNALA_NONRISPONDE, value)
                Me.DoChanged("SegnalaNonRisponde", value, Not value)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta un valore booleano che indica se il sistema deve segnalare la programmazioni du un ricontatto oltre il limite di MinutiRicontatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property SegnalaRicontattiTroppoLontani As Boolean
            Get
                Return TestFlag(Me.m_Flags, CRMFlags.SEGNALA_RICONTATTITROPPOLONTANI)
            End Get
            Set(value As Boolean)
                If (Me.SegnalaRicontattiTroppoLontani = value) Then Exit Property
                Me.m_Flags = SetFlag(Me.m_Flags, CRMFlags.SEGNALA_RICONTATTITROPPOLONTANI, value)
                Me.DoChanged("SegnalaRicontattiTroppoLontani", value, Not value)
            End Set
        End Property


        ''' <summary>
        ''' Restituisce o imposta un valore booleano che indica se il sistema deve segnalare la programmazioni du un ricontatto oltre il limite di MinutiRicontatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property InviaStatisticheGiornaliere As Boolean
            Get
                Return TestFlag(Me.m_Flags, CRMFlags.INVIA_STATISTICHE_GIORNALIERE)
            End Get
            Set(value As Boolean)
                If (Me.InviaStatisticheGiornaliere = value) Then Exit Property
                Me.m_Flags = SetFlag(Me.m_Flags, CRMFlags.INVIA_STATISTICHE_GIORNALIERE, value)
                Me.DoChanged("InviaStatisticheGiornaliere", value, Not value)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta l'indirizzo e-mail utilizzato per inviare le statistiche e gli alert del CRM
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property FromAddress As String
            Get
                Return Me.m_FromAddress
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_FromAddress
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_FromAddress = value
                Me.DoChanged("FromAddress", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il nome visualizzato come mittente
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property FromDisplayName As String
            Get
                Return Me.m_FromDisplayName
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_FromDisplayName
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_FromDisplayName = value
                Me.DoChanged("FromDisplayName", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta l'indirizzo e-mail a cui inviare le segnalazioni e le statistiche
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property TargetAddress As String
            Get
                Return Me.m_TargetAddress
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_TargetAddress
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_TargetAddress = value
                Me.DoChanged("TargetAddress ", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta flags aggiuntivi
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Flags As CRMFlags
            Get
                Return Me.m_Flags
            End Get
            Set(value As CRMFlags)
                Dim oldValue As CRMFlags = Me.m_Flags
                If (oldValue = value) Then Exit Property
                Me.m_Flags = value
                Me.DoChanged("Flags", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta i minuti oltre i quali il nuovo ricontatto programmato deve essere segnalato
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property MinutiRicontatto As Integer
            Get
                Return Me.m_MinutiRicontatto
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_MinutiRicontatto
                If (oldValue = value) Then Exit Property
                Me.m_MinutiRicontatto = value
                Me.DoChanged("MinutiRicontatto", value, oldValue)
            End Set
        End Property


        Protected Overrides Function GetConnection() As CDBConnection
            Return APPConn
        End Function

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_CRMConfig"
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_FromAddress = reader.Read("FromAddress", Me.m_FromAddress)
            Me.m_FromDisplayName = reader.Read("FromDisplayName", Me.m_FromDisplayName)
            Me.m_Flags = reader.Read("Flags", Me.m_Flags)
            Me.m_TargetAddress = reader.Read("TargetAddress", Me.m_TargetAddress)
            Me.m_MinutiRicontatto = reader.Read("MinutiRicontatto", Me.m_MinutiRicontatto)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("FromAddress", Me.m_FromAddress)
            writer.Write("FromDisplayName", Me.m_FromDisplayName)
            writer.Write("Flags", Me.m_Flags)
            writer.Write("TargetAddress", Me.m_TargetAddress)
            writer.Write("MinutiRicontatto", Me.m_MinutiRicontatto)

            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("FromAddress", Me.m_FromAddress)
            writer.WriteAttribute("FromDisplayName", Me.m_FromDisplayName)
            writer.WriteAttribute("Flags", Me.m_Flags)
            writer.WriteAttribute("TargetAddress", Me.m_TargetAddress)
            writer.WriteAttribute("MinutiRicontatto", Me.m_MinutiRicontatto)

            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "FromAddress" : Me.m_FromAddress = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "FromDisplayName" : Me.m_FromDisplayName = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Flags" : Me.m_Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "TargetAddress" : Me.m_TargetAddress = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "MinutiRicontatto" : Me.m_MinutiRicontatto = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select

        End Sub

        Public Sub Load()
            Dim dbSQL As String = "SELECT * FROM [" & Me.GetTableName & "] ORDER BY [ID] ASC"
            Dim dbRis As System.Data.IDataReader = Me.GetConnection.ExecuteReader(dbSQL)
            If (dbRis.Read) Then
                Me.GetConnection.Load(Me, dbRis)
            End If
            dbRis.Dispose()
        End Sub

        Public Shadows Sub Save()
            MyBase.Save()
            CRM.SetConfig(Me)
        End Sub

    End Class


End Class